File: src/dp/knapsack/km46-perf.py
itemNum, totalVolume = map(int, input().strip().split(" "))

itemVolumes = list(map(int, input().strip().split(" ")))
itemValues = list(map(int, input().strip().split(" ")))

f = [0] * (totalVolume + 1)

for i in range(1, itemNum + 1):
    for j in range(totalVolume, 0, -1):
        if j < itemVolumes[i - 1]:
            continue

        f[j] = max(f[j], f[j - itemVolumes[i - 1]] + itemValues[i - itemVolumes[i - 1]])

print(f[totalVolume])

File: src/dp/knapsack/km46-perf.cc
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

int main() {
  int itemNum, totalVolume;
  cin >> itemNum >> totalVolume;

  auto itemVolumes = vector<int>(itemNum, 0);
  auto itemValues = vector<int>(itemNum, 0);

  for (auto i = 0; i < itemNum; i++) {
    cin >> itemVolumes[i];
  }

  for (auto i = 0; i < itemNum; i++) {
    cin >> itemValues[i];
  }

  auto f = vector<int>(totalVolume + 1, 0);

  for (auto i = 1; i <= itemNum; i++) {
    // 倒序
    for (auto j = totalVolume; j >= 1; j--) {
      if (j < itemVolumes[i - 1]) {
        continue;
      }

      f[j] = max(f[j], f[j - itemVolumes[i - 1]] + itemValues[i - 1]);
    }
  }

  cout << f[totalVolume] << endl;
}
